### PR TITLE
dataflow,coord: add reachability introspection views

### PR DESCRIPTION
**Pull request by aljoscha**
Right now, reachability logging as active when the other timely logging tables are active. Is that something we want?

I had one question in there about what I can set as a foreign key.

Also, not sure who would review but @frankmcsherry and @wangandi reviewed similar changes last time.

With this you can create views:
```
> CREATE TABLE source_table (a STRING)

> CREATE VIEW addresses AS
  SELECT mdo.id, mdo.name, mdo.worker, mdoa.address
  FROM mz_dataflow_operator_dataflows AS mdo, mz_dataflow_operator_addresses AS mdoa
  WHERE mdo.dataflow_name LIKE '%public.source_table%' AND mdoa.id = mdo.id AND mdo.worker = mdoa.worker

> CREATE VIEW reachability AS
  SELECT ops.id, ops.name, ops.worker, ops.address, dor.port, dor.update_type, dor.timestamp, dor.count
  FROM addresses ops, mz_dataflow_operator_reachability dor
  WHERE ops.address = dor.address AND ops.worker = dor.worker
```

And then query/tail them:
```
materialize=> select * from addresses;
 id  |                         name                          | worker | address
-----+-------------------------------------------------------+--------+----------
 822 | Map                                                   |      0 | {29,1,7}
 812 | Empty                                                 |      0 | {29,1,2}
 813 | MapInPlace                                            |      0 | {29,1,3}
 819 | Concatenate                                           |      0 | {29,1,6}
 811 | UnorderedInput                                        |      0 | {29,1,1}
 815 | FormArrangementKey                                    |      0 | {29,1,4}
 817 | ArrangeBy[[Column(0)]]                                |      0 | {29,1,5}
 824 | ArrangeBy[[Column(0)]]-errors                         |      0 | {29,1,8}
 810 | Dataflow: materialize.public.source_table_primary_idx |      0 | {29}
 827 | Dataflow: materialize.public.source_table_primary_idx |      0 | {29,1}
(10 rows)

materialize=> select * from reachability;
 id  |      name      | worker | address  | port | update_type |   timestamp   | count
-----+----------------+--------+----------+------+-------------+---------------+-------
 811 | UnorderedInput |      0 | {29,1,1} |    0 | source      | 1628171347541 |     1
(1 row)

materialize=> COPY (TAIL reachability) TO STDOUT;
1628171184999	1	811	UnorderedInput	0	{29,1,1}	0	source	1628171183838	1
1628171185000	-1	811	UnorderedInput	0	{29,1,1}	0	source	1628171183838	1
1628171185000	1	811	UnorderedInput	0	{29,1,1}	0	source	1628171184850	1
1628171186000	-1	811	UnorderedInput	0	{29,1,1}	0	source	1628171184850	1
1628171186000	1	811	UnorderedInput	0	{29,1,1}	0	source	1628171185863	1
1628171187000	-1	811	UnorderedInput	0	{29,1,1}	0	source	1628171185863	1
1628171187000	1	811	UnorderedInput	0	{29,1,1}	0	source	1628171186875	1
1628171188000	-1	811	UnorderedInput	0	{29,1,1}	0	source	1628171186875	1
1628171188000	1	811	UnorderedInput	0	{29,1,1}	0	source	1628171187888	1
```

### CODE REVIEW

**Review by wangandi — APPROVED** [review]
Looks all right. That said, I don't know much about the original commit in differential.

**Review by frankmcsherry — APPROVED** [review]
Looks good to me. The failures seem probably unrelated?

File: src/dataflow-types/src/logging.rs [hunk 1]
```diff
@@ -237,10 +254,18 @@ impl LogVariant {
...
-            )],
+            LogVariant::Timely(TimelyLog::Addresses) => vec![
+                (
+                    LogVariant::Timely(TimelyLog::Operates),
+                    vec![(0, 0), (1, 1)],
+                ),
+                (
+                    // TODO(aljoscha) can this be a foreign key even though (2, 0)
+                    // is not a primary key on Reachability?
+                    LogVariant::Timely(TimelyLog::Reachability),
```
[the remark this answers is not in the document]
Comment (aljoscha) [reply src/dataflow-types/src/logging.rs]: I removed the foreign key declarations. Also, Addresses is not unique on [1, 2], so I also can't add it the other way around. If we ever find out we need it we can add the declarations later, I think.

**Comment by aljoscha** [comment]
@frankmcsherry The test failure was related. There is a persistence test that needs to change whenever builtin tables are added, because it has a view that includes all the builtin arrangements: https://github.com/MaterializeInc/materialize/blob/859a89dd7c4ccf610acc76211a0082f16ac00f02/src/materialized/tests/server.rs#L90

@danhhz or @ruchirK Should I maybe try and change this test to only work on user-defined views, so we don't have to change it again in the future?

**Comment by aljoscha** [comment]
Ah sorry, I stupidly assumed that the test had to do with the new persistence work without checking the context. Nevermind.

**Comment by aljoscha** [comment]
@sploiselle Are you OK with the test changes (it's the second commit)? I'm happy to remove the commit if there's a reason behind also checking for the builtin views.

**Comment by sploiselle** [comment]
@aljoscha We don't have any other tests that ensure system view persistence AFAIK, so my inclination is that we should keep looking at all of the catalog's views. Does the presence of the system views impose something onerous on this PR?

**Comment by sploiselle** [comment]
@aljoscha Think my last comment was totally wrong; removing system views from `test_persistence` is totally fine. We substantiate the views' presence in [testdrive](https://github.com/MaterializeInc/materialize/blob/fdab2f413ee8354b70bc5e27aab5f88e8d1e8e7b/test/testdrive/catalog.td#L296-L315), so this test is redundant.

**Comment by aljoscha** [comment]
Thanks, @sploiselle! (And thanks for double-checking :smiley:)

**Pull request merged**